From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 31 Dec 2023 13:48:55 +0100
Subject: [PATCH] Client perspective - Block state visualization - Apply to
 chunk packets - ClientboundLevelChunkPacketData has a reference to its
 ClientboundLevelChunkWithLightPacket

In `ClientboundLevelChunkPacketData`, adds a reference to its containing `ClientboundLevelChunkWithLightPacket`.
This will later allow us to implement applying the visualization in `ClientboundLevelChunkPacketData`, and then mark the packet as ready when it has finished.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index ccba692507b554a47af5a0dcdd19ed4af57fcf1d..0c427250ff367dbd340c385b1275b8acb0bd1b8e 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -21,6 +21,12 @@ import net.minecraft.world.level.levelgen.Heightmap;
 
 public class ClientboundLevelChunkPacketData {
     private static final int TWO_MEGABYTES = 2097152;
+    // Fiddle start - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
+    /**
+     * The {@link ClientboundLevelChunkWithLightPacket} that contains this instance as its chunk data.
+     */
+    private final ClientboundLevelChunkWithLightPacket containingPacket;
+    // Fiddle end - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
     private final CompoundTag heightmaps;
     private final byte[] buffer;
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
@@ -34,7 +40,10 @@ public class ClientboundLevelChunkPacketData {
     // Paper end
 
     // Paper start - Anti-Xray - Add chunk packet info
-    public ClientboundLevelChunkPacketData(LevelChunk chunk, @Nullable com.destroystokyo.paper.antixray.ChunkPacketInfoAntiXray chunkPacketInfo) { // Fiddle - client perspective - block state visualization - apply to chunk packets - every ChunkPacketInfo is ChunkPacketInfoAntiXray - the parameter type is always ChunkPacketInfoAntiXray, and is non-null if anti-x-ray is to be applied (but if it is non-null, still no anti-x-ray would be applied if the used ChunkPacketBlockController never applies anti-x-ray)
+    // Fiddle start - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
+    public ClientboundLevelChunkPacketData(ClientboundLevelChunkWithLightPacket containingPacket, LevelChunk chunk, @Nullable com.destroystokyo.paper.antixray.ChunkPacketInfoAntiXray chunkPacketInfo) { // Fiddle - client perspective - block state visualization - apply to chunk packets - every ChunkPacketInfo is ChunkPacketInfoAntiXray - the parameter type is always ChunkPacketInfoAntiXray, and is non-null if anti-x-ray is to be applied (but if it is non-null, still no anti-x-ray would be applied if the used ChunkPacketBlockController never applies anti-x-ray)
+        this.containingPacket = containingPacket;
+        // Fiddle end - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
         // Paper end
         this.heightmaps = new CompoundTag();
 
@@ -71,7 +80,10 @@ public class ClientboundLevelChunkPacketData {
 
     }
 
-    public ClientboundLevelChunkPacketData(FriendlyByteBuf buf, int x, int z) {
+    // Fiddle start - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
+    public ClientboundLevelChunkPacketData(ClientboundLevelChunkWithLightPacket containingPacket, FriendlyByteBuf buf, int x, int z) {
+        this.containingPacket = containingPacket;
+        // Fiddle end - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
         this.heightmaps = buf.readNbt();
         if (this.heightmaps == null) {
             throw new RuntimeException("Can't read heightmap in packet for [" + x + ", " + z + "]");
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
index 1366eed8169ce9466c125e1547ba008e298707d8..6f892646c952a055a8947b14f6fc8735aae33e6e 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundLevelChunkWithLightPacket.java
@@ -37,7 +37,7 @@ public class ClientboundLevelChunkWithLightPacket extends org.fiddlemc.fiddle.pa
         boolean isChunkPacketInfoNeeded = modifyBlocks;
         var chunkPacketInfo = isChunkPacketInfoNeeded ? chunk.getLevel().chunkPacketBlockController.getChunkPacketInfo(this, chunk, modifyBlocks) : null; // No effective changes to this field except the type is inferred
         // Fiddle end - client perspective - block state visualization - apply to chunk packets - every ChunkPacketInfo is ChunkPacketInfoAntiXray
-        this.chunkData = new ClientboundLevelChunkPacketData(chunk, chunkPacketInfo);
+        this.chunkData = new ClientboundLevelChunkPacketData(this, chunk, chunkPacketInfo); // Fiddle - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
         // Paper end
         this.lightData = new ClientboundLightUpdatePacketData(chunkPos, lightProvider, skyBits, blockBits);
         chunk.getLevel().chunkPacketBlockController.modifyBlocks(this, chunkPacketInfo); // Paper - Anti-Xray - Modify blocks
@@ -47,7 +47,7 @@ public class ClientboundLevelChunkWithLightPacket extends org.fiddlemc.fiddle.pa
         super(); // Fiddle - client perspective - block state visualization - apply to chunk packets - store connection explicitly
         this.x = buf.readInt();
         this.z = buf.readInt();
-        this.chunkData = new ClientboundLevelChunkPacketData(buf, this.x, this.z);
+        this.chunkData = new ClientboundLevelChunkPacketData(this, buf, this.x, this.z); // Fiddle - client perspective - block state visualization - apply to chunk packets - ClientboundLevelChunkPacketData has a reference to its ClientboundLevelChunkWithLightPacket
         this.lightData = new ClientboundLightUpdatePacketData(buf, this.x, this.z);
     }
 
